content/post/review-building-static-analysis-tools.md

```diff
@@ -0,0 +1,249 @@
+
+[**Infer static analysis tool**](http://fbinfer.com/)
+TK does hugo have nice big links?
+
+The reasoning behind why Google is not interested in more complex analysis at this point are explicitly listed in the paper:
+
+* *Large investment.* Although theoretically better and more complex analysis exist, it will require non-trivial engineering effort to scale
+
+* *High upfront cost.* It’s very hard to get the data, even to just estimate the cost-benefit ratio for such improvements, to justify research/implementation efforts
+
+* *Work still would be needed to reduce false-positive rates*
+
+* *Finding bugs is easy.* When a codebase is large enough, it will contain practically any imaginable code patter. Many more low-hanging fruits exist using simpler approaches
+
+I find it to be particularly curious, as none of these conditions are set in stone:
+
+* data is hard to get without decrease in productivity of engineers inside the company, due to the way false-positives are measured (more on it in the next section)
+
+* what seems not feasible now, may become feasible later with more computation power and i.e in cases when some analysis could be
+ cast as some kind of optimization problem
+
+* and many other projects and companies could benefit from more advanced analysis, where it is feasible conduct it due to codebases of smaller sizes
+
+Main critique of practical usefulness of any automated analysis approach is based on existence of *false-positives results*: issues reported as a defects, that either could not happen or are not relevant to the recipients of analysis.
+
+Traditionally, in software industry two main types of “analyzers” could be vaguely distinguished and have proved to be useful:
+ — *Style checkers* or [Linters](https://en.wikipedia.org/wiki/Lint_(software)) (i.e [CheckStyle](https://github.com/checkstyle/checkstyle), [Pylint](https://www.pylint.org/), [Golint](https://github.com/golang/lint), ets)
+ — *Bug-finding tools*, that may extend the compiler ([FindBugs](https://github.com/findbugsproject/findbugs), [Error-Prone](http://errorprone.info/), [Clang-Tidy](http://clang.llvm.org/extra/clang-tidy), [ClangThreadSafety Analysis](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html), [Govet](https://golang.org/cmd/vet/), [CheckerFramework](https://checkerframework.org), etc)
+
+For this kind of tools, different integration points in a software development process of an organization exist, where results could be presented to the developers. They are discussed in the paper in the chronological order of BugBot team’s attempts of integrating an existing OSS static analysis tool for Java — [FindBugs](https://github.com/findbugsproject/findbugs) into the Google’s developer workflow.
```

Comment:
Different integration points in the software development process of an organization exist for this kind of tools where **the** results **can** be presented

Reply:
of the BugBot team's attempts ~of~ **to** integrate